_posts/2018-02-12-Reusable-features-in-React-Redux.md

```diff
@@ -0,0 +1,188 @@
+       TodoListItem.js
+     actions.js
+     selectors.js 
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```
+
+
+It is always good to break features into smaller sub-features with encapsulated logic to reduce the complexity.
+Try to think of it as applying Single Responsibility Principle. One feature does only one thing and
+it's entire logic is isolated in it's directory. As you applications grows and you add more and more features, 
+it really pays of to structure your code using these rules. Somebody that will come to work on a project after
+you will really thank you ;]
+
+The standard feature has unfortunately one disadvantage. It is statically mounted on specific place in a state. 
+Imagine a situation when you want to display two Todo Lists on the same page.
+If you do that, it will work and the Todo Lists will display. The problem arises when you try to interact with them. 
+Changes done in one list (checking the item as done, reorder...) will be reproduced also on the second one.
+The cause of this effect is the fact that they are both focused on the same place in the Redux state and 
+act accordingly on Redux state changes. You may never intercept this situation, it really depends
+on the type of the application you are building and the use-cases. But if you do, there is a solution.
+Enter the realm of **reusable features**.
+
+## Reusable features
+
+[Some people](https://kickstarter.engineering/namespacing-actions-for-redux-d9b55a88b1b1) has already dealt with this problem. My solution is a little bit different but
+at the end we're trying to solve the same thing. So what is the **reusable feature** ?
+It's basically a standard feature but instead of things defined statically, everything (except components)
+is defined inside **factory functions**. Factory functions allows us to create things in lazy configurable manner.
+Observe e.g. the difference between standard feature actions and reusable one.
+
+
+#### Standard feature actions.js
+
+```js
+const STANDARD_FEATURE_INCREMENT_CLICK = `${namespace}_STANDARD_FEATURE_INCREMENT_CLICK`;
+
+export const incrementClick = createAction(STANDARD_FEATURE_INCREMENT_CLICK);
+```
+     
+#### Reusable feature actions.js
+
+```js
+export default function createActions(namespace, selectors) {
+
+  const REUSABLE_FEATURE_INCREMENT_CLICK = `${namespace}_REUSABLE_FEATURE_INCREMENT_CLICK`;
+
+  const incrementClick = createAction(REUSABLE_FEATURE_INCREMENT_CLICK);
+
+
+  return {
+    REUSABLE_FEATURE_INCREMENT_CLICK: REUSABLE_FEATURE_INCREMENT_CLICK,
+
+    incrementClick,
+  };
+}
+```
+
+Yeah, and that's it. You define everything through the factory functions. 
+At this stage the reusable feature acts as a mere template. Check how the full
+implementation of [reusable feature looks](https://github.com/char0n/react-reusable-features/tree/master/src/app/reusable-feature) in our [reference repository](https://github.com/char0n/react-reusable-features).
```

Comment:
Either:

"implementation of a reusable feature looks"

-OR-

"implementation of reusable features looks"

("a feature" vs "features")